cannon: Create framework for differential tests

mbaxter opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

Introduce a new "DiffTester" framework to run differential tests for single vm steps.  This allows us to inspect test expectations and inject additional test cases as necessary.  

For example, if a test expects the vm's memory to be updated, we can [inject an additional test case](https://github.com/ethereum-optimism/optimism/blob/439cfb760d612fa4c3e5fc9f7e27b86e03e753ce/cannon/mipsevm/tests/difftester.go#L116-L162) to ensure that any overlapping memory reservation is cleared.  This ensures that memory reservations are cleared as expected when any instruction or syscall writes to memory.  

You can see an example of the new framework being used [here](https://github.com/ethereum-optimism/optimism/blob/439cfb760d612fa4c3e5fc9f7e27b86e03e753ce/cannon/mipsevm/tests/testfuncs_test.go#L174-L201).

**Follow-up Work**

Ultimately we should migrate all Cannon differential tests to use this new framework.  Remaining tests will be migrated in follow-up PRs. 

**Metadata**

Part of https://github.com/ethereum-optimism/optimism/issues/16483

mbaxter committed:
Unify ExpecteMemoryWrite function calls

mbaxter committed:
Add expectation helper ExpectMemoryReservationCleared()

mbaxter committed:
In progress - start working on difftester framework

mbaxter committed:
Use factory methods for test dependencies

mbaxter committed:
For simplicity, drop validator logic, TestContext for the time being

mbaxter committed:
Rework difftester - add test cases intead of modifying tests

mbaxter committed:
Drop ExpectedState wrapper

mbaxter committed:
Drop state wrapper

mbaxter committed:
Rework difftester API

Inphi commented on cannon/mipsevm/tests/difftester.go:
```diff
@@ -0,0 +1,201 @@
+	return d
+}
+
+func (d *DiffTester[T]) Run(t *testing.T, testCases []T, randSeed int64, opts ...TestOption) {
```
nit: The Run function could be simplified for callers. All inputs except maybe `t`, so that a `DiffTester` can be reused, could be preconfigured. Similar to how `setExpectations` is configured.
The seed could the have a default value if it's not configured.

mbaxter replied:
I think the current `Run` signature holds all of the values I wouldn't expect to be shared across test runs, so that we could reuse the DiffTester (which just remembers initialization / expectation logic).  And I kind of like forcing the caller to set a random seed so we get some variety across tests even though it's a little verbose.  

But happy to iterate / discuss more.

Inphi replied:
what do you think of mixing the test name with a fixed seed? To keep the inputs varied while  making it slightly easier to use and read the test setup.

mbaxter replied:
I like that idea.  I've implemented this [here](https://github.com/ethereum-optimism/optimism/pull/16608/commits/4bb9a075bc5d74e1c66b7f0ab063ec84243df41b) by just hashing the testcase string.  WDYT?

Inphi reviewed (approved):
Looks good. I can see difftester being useful for enforcing other state properties in the future.

mbaxter committed:
Rework TestNamer handling

mbaxter committed:
Expand DiffTester to accommodate more types of tests

- Pass the vm to SetExpectationsFn
- Return an ExpectedExecResult from SetExpectationsFn to handle panics

mbaxter committed:
Migrate some tests that use the new DiffTester features

mbaxter committed:
Add some unit tests for difftester

mbaxter committed:
Cleanup - reorganize helpers

mbaxter committed:
Cleanup - narrow test assertions, drop unused helper

mbaxter committed:
Merge branch 'develop' into mbax/issue-16483/add-difftester-framework

mbaxter committed:
Run tests in parallel

Inphi commented on cannon/mipsevm/tests/difftester_test.go:
```diff
@@ -0,0 +1,364 @@
+}
+
+func (m *mockTestRunner) Run(name string, fn testFn) bool {
+	// TODO: How should we handle panics?
```
Since panics in the VM are explicitly caught by tests, a test shouldn't panic. Else it's a buggy test and I'd like that to be very visible with a stack trace, rather than result in a "failed" test.

mbaxter replied:
Ah, this is a stale TODO.  I have some unit tests [here](https://github.com/ethereum-optimism/optimism/blob/d2b7eb5c60c0b994a5958083a18f50533141a28a/cannon/mipsevm/tests/difftester_test.go#L170-L174) that assert the vm should not panic when it actually does panic. This means that a panic is emitted when running these tests, so have to catch it in this test helper here. 

The difftester itself does not try handle panics.  So if a test panics, we'll see it 👍

mbaxter committed:
Auto-generate random seed from testcase name

mbaxter committed:
Cut todo
